### PR TITLE
Add batching manager

### PR DESCRIPTION
**Pull request by abocquet**
Paves the way for proper handling of the batching in https://github.com/pierreguilmin/torchqdynamics/pull/58. Indeed, when we add time dependent operators, following the dimensions becomes necessary, hence the `BatchingHandler` class

It'll also avoid duplicate code for next numerical schemas (e.g SMESolve)

### CODE REVIEW

File: torchqdynamics/batching.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from typing import Tuple
+
+from torch import Tensor
+
+from .utils.tensor_types import OperatorLike, TDOperatorLike, to_tensor
+from .utils.utils import is_ket, ket_to_dm
+
+
+class BatchingHandler:
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply torchqdynamics/batching.py:9]: Also the file name should match the class name.

File: torchqdynamics/batching.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        self.dtype = dtype
+        self.device = device
+
+        self.H_is_batched = False
+        self.state_is_batched = False
+
+    def batch_H_and_state(
+        self, H: TDOperatorLike, state: OperatorLike, state_to_dm=False
+    ) -> Tuple[Tensor, Tensor]:
+        """Batches H and the state (psi or rho)"""
```
Comment (pierreguilmin) [inline torchqdynamics/batching.py:20]:
```suggestion
        """Batches H and the state (psi or rho)."""
```

File: torchqdynamics/batching.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply torchqdynamics/batching.py:9]: And then we can add the `t_save` conversion as well.

File: torchqdynamics/batching.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+class BatchingHandler:
+    def __init__(self, dtype, device):
+        self.dtype = dtype
+        self.device = device
+
+        self.H_is_batched = False
+        self.state_is_batched = False
+
+    def batch_H_and_state(
+        self, H: TDOperatorLike, state: OperatorLike, state_to_dm=False
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply torchqdynamics/batching.py:18]: Also fix the typing of `__init__` and `unbatch`.

**Review by gautierronan — COMMENTED** [review]
Really nice PR @abocquet. I love it ! I don't have any more comments than what @pierreguilmin already proposed :)

File: torchqdynamics/mesolve/mesolve.py [hunk 4]
```diff
@@ -153,11 +137,7 @@ def mesolve(
...
     # get saved tensors and restore correct batching
     rho_save, exp_save = solver.y_save, solver.exp_save
-    if rho0.ndim == 2:
-        rho_save = rho_save.squeeze(1)
-        exp_save = exp_save.squeeze(1)
-    if H.ndim == 2:
-        rho_save = rho_save.squeeze(0)
-        exp_save = exp_save.squeeze(0)
+    rho_save = formatter.unbatch(rho_save)
+    exp_save = formatter.unbatch(exp_save)
```
Comment (gautierronan) [inline torchqdynamics/mesolve/mesolve.py:141]: Found a small bug (already present in the previous code, but it's a good PR to fix it). If no `exp_ops` are passed, then `exp_save` is None, and this raises an error. So we need to check if `exp_save` is not None somewhere.

Same thing for `sesolve`.

File: torchqdynamics/batching.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abocquet) [reply torchqdynamics/batching.py:9]: But `t_save` isn't batched is it?

File: torchqdynamics/tensor_formatter.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+from typing import Optional, Tuple
+
+from torch import Tensor
+
+from .utils.tensor_types import OperatorLike, TDOperatorLike, to_tensor
+from .utils.utils import is_ket, ket_to_dm
+
+
+class TensorFormatter:
+    def __init__(self, dtype, device):
```
Comment (pierreguilmin) [inline torchqdynamics/tensor_formatter.py:10]: Typing missing.

File: torchqdynamics/tensor_formatter.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        return H_batched, state_batched
+
+    def batch(self, operator: OperatorLike) -> Tensor:
+        operator = to_tensor(
+            operator, dtype=self.dtype, device=self.device, is_complex=True
+        )
+        return operator[None, ...] if operator.ndim == 2 else operator
+
+    def unbatch(self, save: Optional[Tensor]):
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply torchqdynamics/tensor_formatter.py:52]: Please use `Tensor | None`.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Nitpicking.

File: torchqdynamics/tensor_formatter.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        return H_batched, state_batched
+
+    def batch(self, operator: OperatorLike) -> Tensor:
+        operator = to_tensor(
+            operator, dtype=self.dtype, device=self.device, is_complex=True
+        )
+        return operator[None, ...] if operator.ndim == 2 else operator
+
+    def unbatch(self, save: Tensor | None):
```
Comment (pierreguilmin) [inline torchqdynamics/tensor_formatter.py:57]: Also
```suggestion
    def unbatch(self, save: Tensor | None) -> Tensor | None:
```

**Comment by pierreguilmin** [comment]
You need a `from __future__ import annotations` to fix the CI.

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**